nibabel/cifti2/parse_cifti2_fast.py

```diff
@@ -0,0 +1,467 @@
+        self._content = parser.header
+        return self._content
+
+    def _mangle(self, value):
+        if not isinstance(value, Cifti2Header):
+            raise ValueError('Can only mangle a Cifti2Header.')
+        return value.to_xml()
+
+extension_codes.add_codes((
+    (Cifti2Extension.code, 'cifti', Cifti2Extension),))
+
+intent_codes.add_codes((
+    # The codes below appear on the CIFTI-2 standard
+    # http://www.nitrc.org/plugins/mwiki/index.php/cifti:ConnectivityMatrixFileFormats
+    # https://www.nitrc.org/forum/attachment.php?attachid=341&group_id=454&forum_id=1955
+    (3000, 'dense fiber/fan samples', (), 'NIFTI_INTENT_CONNECTIVITY_UNKNOWN'),
+    (3001, 'dense connectivity', (), 'NIFTI_INTENT_CONNECTIVITY_DENSE'),
+    (3002, 'dense data series/fiber fans', (),
+     'NIFTI_INTENT_CONNECTIVITY_DENSE_SERIES'),
+    (3003, 'parcellated connectivity', (),
+     'NIFTI_INTENT_CONNECTIVITY_PARCELLATED'),
+    (3004, 'parcellated data series', (),
+     "NIFTI_INTENT_CONNECTIVITY_PARCELLATED_SERIES"),
+    (3006, 'dense scalar', (),
+     'NIFTI_INTENT_CONNECTIVITY_DENSE_SCALARS'),
+    (3007, 'dense label', (),
+     'NIFTI_INTENT_CONNECTIVITY_DENSE_LABELS'),
+    (3008, 'parcellated scalar', (),
+     'NIFTI_INTENT_CONNECTIVITY_PARCELLATED_SCALAR'),
+    (3009, 'parcellated dense connectivity', (),
+     'NIFTI_INTENT_CONNECTIVITY_PARCELLATED_DENSE'),
+    (3010, 'dense parcellated connectivity', (),
+     'NIFTI_INTENT_CONNECTIVITY_DENSE_PARCELLATED'),
+    (3011, 'parcellated connectivity series', (),
+     'NIFTI_INTENT_CONNECTIVITY_PARCELLATED_PARCELLATED_SERIES'),
+    (3012, 'parcellated connectivity scalar', (),
+     'NIFTI_INTENT_CONNECTIVITY_PARCELLATED_PARCELLATED_SCALAR')))
+
+
+class _Cifti2AsNiftiHeader(Nifti2Header):
+    ''' Class for Cifti2 header extension '''
+
+    @classmethod
+    def may_contain_header(klass, binaryblock):
+        if not Nifti2Header.may_contain_header(binaryblock):
+            return False
+        hdr = Nifti2Header(binaryblock=binaryblock[:Nifti2Header.sizeof_hdr])
+        intent_code = hdr.get_intent('code')[0]
+        return intent_code >= 3000 and intent_code < 3100  # and intent_code != 3002
+
+
+class _Cifti2AsNiftiImage(Nifti2Image):
+    header_class = _Cifti2AsNiftiHeader
+    files_types = (('image', '.nii'),)
+    valid_exts = ('.nii',)
+    makeable = False
+    rw = True
+
+    def __init__(self, dataobj, affine, header=None,
+                 extra=None, file_map=None):
+        """Convert NIFTI-2 file to CIFTI2"""
+        super(_Cifti2AsNiftiImage, self).__init__(dataobj=dataobj,
+                                                  affine=affine,
+                                                  header=header,
+                                                  extra=extra,
+                                                  file_map=file_map)
+
+        # Get cifti header from extension
+        self.cifti_img = reduce(lambda accum, newval: newval
```

Comment:
There resubmitted the corrected code. It's just a linear search and the semantics are the same for either the first or last element found as there should be just one.